src/amd/compiler/aco_util.h

```diff
@@ -0,0 +1,235 @@
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ *
+ */
+
+#ifndef ACO_UTIL_H
+#define ACO_UTIL_H
+
+#include <utility>
+
+namespace aco {
+
+/*!	\brief		Definition of a span object
+*
+* \details		A "span" is an "array view" type for holding a view of a contiguous data
+*				array. the "span" object does not and cannot own such data itself. A span
+*				provides an easy way to provide overloads for anything operating on
+*				contiguous sequences without needing to manage pointers and array lengths
+*				manually.
+
+*				A span is conceptually a pointer (ptr) and a length (size) into an already
+*				existing array of contiguous memory; the array it represents references the
+*				elements "ptr[0] .. ptr[size-1]". Passing a properly-constructed "span"
+*				instead of raw pointers avoids many issues related to index out of bounds
+*				errors.
+*/
+template <typename T>
+class span {
+public:
+	using value_type             = T;
+	using pointer                = value_type*;
+	using const_pointer          = const value_type*;
+	using reference              = value_type&;
+	using const_reference        = const value_type&;
+	using iterator				 = pointer;
+	using const_iterator         = const_pointer;
+	using reverse_iterator		 = std::reverse_iterator<iterator>;
+	using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+	using size_type              = std::size_t;
+	using difference_type        = std::ptrdiff_t;
+
+	/*!	\brief						Compiler generated default constructor
+	*/
+	constexpr span() = default;
+
+	/*!	\brief						Constructor taking a pointer and the length of the span
+	*	\param[in]	data			Pointer to the underlying data array
+	*	\param[in]	length			The size of the span
+	*/
+	constexpr span(pointer data, const size_type length)
+	    : data{ data } , length{ length } {}
+
+	/*!	\brief						Returns an iterator to the begin of the span
+	*	\return						data
+	*/
+	iterator begin() noexcept {
```

Comment:
can't these all also be declared `constexpr`?

Reply:
Unfortunately this is not possible in C++11 as that does not allow overloads of `constexpr`. Consequently there is a clash between the `const` and non-`const` version of `begin`

Reply:
Right. I guess that'll have to wait until eventual move to C++14.

Reply:
Yeah, although the C++11 requirement is kind of strange given the requirement on bleeding edge LLVM